### PR TITLE
Refactored pkg/scheduler/utils to podinfo and heap packages.

### LINKED ISSUE #71863: [scheduler cleanup phase 2]: Rename `pkg/scheduler/util` to `pkg/scheduler/internal/podinfo`
**PLEASE COORDINATE WITH SIG-SCHEDULING BEFORE WORKING ON THIS ISSUE.** We want to avoid the situation where two people accidentally start work on the same fix.

Main tracking issue: https://github.com/kubernetes/kubernetes/issues/68951

If you contributed in phase 1, please give others a chance before claiming this issue. You can also help out by reviewing the phase 2 PRs.

This package contains only functions related to pods. Having “util” as a name is generally discouraged, since it doesn’t tell readers anything about what is in that package. This change will also mirror the change to create `pkg/scheduler/nodeinfo` which will have a bunch of node-related types and functions.

`GetPodPriority` from this package is used by `pkg/kubelet/eviction`. In order to make our package internal, we should do one of the following. (1 is my preference)

1. Copy the function into kubelet and remove the dependency
2. Move the function into a different area like `apimachinery` or `client-go`
3. Make the funcion an official part of the scheduler’s public go libraries

/sig scheduling
/kind cleanup
/help
/good-first-issue

### PR DESCRIPTION
**Pull request by hprateek43**
**What type of PR is this?**
kind cleanup

**What this PR does / why we need it:**
#71863 is an active open request for scheduler cleanup. I have picked up that request and moved the generic util package into more meaningful internal packages podinfo and heap. Necessary changes in the scheduler and kubelet eviction packages have been made to support those changes.

Fixes #71863

**Special notes for your reviewer**:
The previous PR used schedulerinternalpodinfo as the package reference.The name was found to be too long by members in the previous thread and has been reduced to just podinfo. Let me know if we need any refactoring to rename this to something more meaningful. Also the package is in /pkg/scheduler/internal/podinfo and /pkg/scheduler/internal/heap, so the word internal in path can be a potential break. Comment if we can move them to more meaningful areas.

```release-note
NONE
```

### CODE REVIEW

**Comment by hprateek43** [comment]
@misterikkit Can you review this PR and let me know the changes ?

**Comment by hprateek43** [comment]
@lavalamp @derekwaynecarr Can you review this request?

**Comment by lavalamp** [comment]
You should get a scheduler-specific review, e.g. @bsalamat or @misterikkit.

I'm not sure why the go.sum files all changed; also, the commits seem arranged sub-optimally (consider using fetch & rebase rather than merge?).

File: go.mod
Comment (liggitt) [inline go.mod]: drop this commit

File: staging/src/k8s.io/api/go.sum
Comment (liggitt) [inline staging/src/k8s.io/api/go.sum]: drop this commit as well

**Comment by liggitt** [comment]
there shouldn't be any go.{sum,mod} changes required from what I can see. I'd suggest squashing/rebasing on master and dropping the changes to those files

**Comment by hprateek43** [comment]
@liggitt Will do so. Should I amend the commit of this PR or open a new PR?

**Comment by hprateek43** [comment]
> there shouldn't be any go.{sum,mod} changes required from what I can see. I'd suggest squashing/rebasing on master and dropping the changes to those files

I had create a new module for podinfo and heap in /pkg/scheduler/internal. Are the changes in go.{mod|sum} due to that?

File: pkg/scheduler/algorithm/predicates/metadata.go
Comment (misterikkit) [inline pkg/scheduler/algorithm/predicates/metadata.go]: Please move this import down to the other `k8s.io` imports. The preferred style is usually,
```
<standard library imports>
<blank line>
<everything else>
```

When I'm writing imports, I like to add one extra level of organization:
```
<standard library imports>
<blank line>
<libraries not part of scheduler>
<blank line>
<scheduler libraries>
```

Comment (misterikkit) [reply pkg/scheduler/algorithm/predicates/metadata.go]: (and please make the same fix in the other files as well.)

File: pkg/scheduler/internal/heap/heap.go
Comment (misterikkit) [inline pkg/scheduler/internal/heap/heap.go]: It feels to me like `LessFunc` belongs in this `heap` package. Is it still used anywhere within the `podinfo` package?

File: pkg/scheduler/internal/podinfo/clock.go
Comment (misterikkit) [inline pkg/scheduler/internal/podinfo/clock.go]: I forgot we had clock stuff in here. I should probably file a separate issue to delete that. :smile:

File: pkg/scheduler/internal/podinfo/utils.go
Comment (misterikkit) [inline pkg/scheduler/internal/podinfo/utils.go]: Please rename this file to `podinfo.go`.

File: go.mod
Comment (misterikkit) [reply go.mod]: Yeah, this shouldn't be included in the PR. Is the change needed at all?

**Review by misterikkit — COMMENTED** [review]
Thanks for sending this cleanup! Except for a few comments, this looks good.

I would prefer if the PR title and the git commit message described the changes being made rather than the issue being fixed.

File: go.mod
Comment (hprateek43) [reply go.mod]: @misterikkit @liggitt Squashed and rebased. Dropped commits as per recommendation

File: staging/src/k8s.io/api/go.sum
Comment (hprateek43) [reply staging/src/k8s.io/api/go.sum]: @liggitt Dropped

File: pkg/scheduler/algorithm/predicates/metadata.go
Comment (hprateek43) [reply pkg/scheduler/algorithm/predicates/metadata.go]: This happened as a result of auto refactoring by GoLand. I have squashed and rebased the commits. Please review the same.

File: pkg/scheduler/internal/podinfo/clock.go
Comment (hprateek43) [reply pkg/scheduler/internal/podinfo/clock.go]: @misterikkit We should delete the clock stuff or should we migrate it into another package ?

File: pkg/scheduler/internal/podinfo/utils.go
Comment (hprateek43) [reply pkg/scheduler/internal/podinfo/utils.go]: utils.go -> podinfo.go
utils_test.go -> podinfo_test.go

Is this the expected rename ?

**Comment by hprateek43** [comment]
@liggitt @misterikkit Do the changes look more aligned now? I have omitted the rename change for now. If the changes look good let us run a test on these, and then I will rebase the commits to introduce the file rename. Just looking to check if something is breaking or not.

**Comment by liggitt** [comment]
changes are now contained to the scheduler package, will defer to @kubernetes/sig-scheduling-pr-reviews for the rest

**Comment by hprateek43** [comment]
@misterikkit Modified the PR title to actual refactoring being done. Let me know if a better PR title can be used.

File: pkg/scheduler/internal/podinfo/clock.go
Comment (misterikkit) [reply pkg/scheduler/internal/podinfo/clock.go]: For this PR, I'd like to keep the clock stuff in the podinfo package.

After this PR, I want to delete it since there is already clock functionality in https://godoc.org/k8s.io/apimachinery/pkg/util/clock.

File: pkg/scheduler/internal/podinfo/utils.go
Comment (misterikkit) [reply pkg/scheduler/internal/podinfo/utils.go]: yes, thanks

File: pkg/kubelet/eviction/helpers.go
Comment (misterikkit) [inline pkg/kubelet/eviction/helpers.go]: s/return/returns/

File: pkg/scheduler/algorithm/predicates/metadata.go
Comment (misterikkit) [reply pkg/scheduler/algorithm/predicates/metadata.go]: Looks like the import is still in the top section. Can you override GoLand's placement? (if you manually move it, does GoLand put it back?)

File: pkg/scheduler/internal/heap/heap.go
Comment (misterikkit) [reply pkg/scheduler/internal/heap/heap.go]: It looks like `LessFunc` is used by both `SortableList` and the heap package. If I had a lot of time on my hands, there are ways I would want to reorganize/refactor this. But I think it's good enough as-is.

**Review by misterikkit — CHANGES_REQUESTED** [review]
We will still need approval for pkg/controller and pkg/kubelet, but let's get scheduler approval first.

Please go ahead with the file renaming and squash commits.

**Comment by hprateek43** [comment]
@misterikkit Went back to basics with vim and changed the import placement locations. I have also renamed the files in the /scheduler/internal/podinfo section.

**Comment by hprateek43** [comment]
@bowei @andrewsykim Can you review this PR?

**Comment by misterikkit** [comment]
/lgtm

looks like you will need a rebase to resolve merge conflicts.

**Comment by hprateek43** [comment]
@misterikkit I am awaiting the test results before I rebase. Will I get them as it seems to take forever and the LGTM label has been removed?

**Comment by misterikkit** [comment]
/ok-to-test

The tests that matter can be run on your local machine without having to
wait forever. Either with `make test` or `bazel test`

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/testing.md

*From:*Kubernetes Prow Robot <notifications@github.com>
*Date:*Fri, May 3, 2019, 7:01 AM
*To:*kubernetes/kubernetes
*Cc:*Jonathan Basseri, Mention

@hprateek43 <https://github.com/hprateek43>: Cannot trigger testing until a

**Comment by misterikkit** [comment]
I chatted with @hprateek43 on May 6 about resolving the merge here. Other reviewers feel free to wait until we get that resolved.

**Comment by hprateek43** [comment]
@misterikkit I was out on travel recently. Will be picking up this today and sending the commit. Just need to handle the test cases now.

**Comment by hprateek43** [comment]
@misterikkit I would probably need you to review the commits and resolve the conflicting files in the commits. I think only users with write access can perform that.

**Comment by hprateek43** [comment]
@misterikkit Can you review and add lgtm ?

**Comment by hprateek43** [comment]
@bowei @andrewsykim Can you review this PR?

File: pkg/kubelet/eviction/helpers.go
Comment (ahg-g) [inline pkg/kubelet/eviction/helpers.go]: can you please keep the order as it was before? those imports should be at the top.

File: pkg/kubelet/eviction/helpers.go
Comment (ahg-g) [inline pkg/kubelet/eviction/helpers.go]: we now have two implementations of this, I think we should have a single implementation under pkg/api/pod/util.go

**Review by ahg-g — COMMENTED** [review]
the packages are still named util, we should change it to heap and podinfo.

File: pkg/kubelet/eviction/helpers.go
Comment (misterikkit) [reply pkg/kubelet/eviction/helpers.go]: This was discussed in a previous (closed) PR, but for something small like this, having two copies is not the end of the world. I support moving it to a neutral library (neither kubelete nor scheduler) but I don't think that should be part of this PR.

File: pkg/controller/.import-restrictions
Comment (misterikkit) [inline pkg/controller/.import-restrictions]: Is this line needed? Was it auto-generated?

File: pkg/kubelet/eviction/helpers.go
Comment (misterikkit) [reply pkg/kubelet/eviction/helpers.go]: +1
I'm surprised lint didn't complain about this.

File: pkg/kubelet/eviction/helpers.go
Comment (ahg-g) [reply pkg/kubelet/eviction/helpers.go]: Not doing it in this PR is fine, but we should move it eventually, even small things like this can diverge in semantics without noticing.

**Review by misterikkit — COMMENTED** [review]
This just needs a couple tweaks before LGTM.

Some of the moved files still have `package util` at the top. Those need to be renamed.

File: pkg/controller/.import-restrictions
Comment (hprateek43) [reply pkg/controller/.import-restrictions]: I think it is autogenerated. Since I refactored util to internal, GoLand changed the dependency there. I am not sure if it is needed, but will check if it is required.

Comment (hprateek43) [reply pkg/controller/.import-restrictions]: We need this change. Just went through the dependencies.

**Comment by hprateek43** [comment]
@misterikkit Can you review the changes now?

**Comment by hprateek43** [comment]
@andrewsykim @bowei @misterikkit Can you review and give a /lgtm ?

**Comment by hprateek43** [comment]
@deads2k @sig-scheduling-maintainers @Random-Liu Can you approve this PR?

File: pkg/controller/.import-restrictions
Comment (deads2k) [inline pkg/controller/.import-restrictions]: This breaks a logical dependency contract.  The controller and scheduler are distinct.  Having the controller depend on an internal package in the scheduler means the two can never be separated.  This is a bad outcome.  We should not make this change.

**Comment by deads2k** [comment]
See https://github.com/kubernetes/kubernetes/pull/76990/files#r294937633 . The controller-manager must not rely on an internal scheduler package.  We should not pursue this pull if it means that internal packages are leaking into other processes.  If those those processes need their dependencies fixed, do that first, then move the package.

/hold

**Comment by deads2k** [comment]
/assign @ravisantoshgudimetla 

have a look to see if you know of anyone already depending on these packages before they move to internal.

File: pkg/controller/.import-restrictions
Comment (deads2k) [reply pkg/controller/.import-restrictions]: > This breaks a logical dependency contract. The controller and scheduler are distinct. Having the controller depend on an internal package in the scheduler means the two can never be separated. This is a bad outcome. We should not make this change.

hmmm.... Tests passed.  I'm wondering if this is a transitive dep and this file just lists all possible transitives you can rely upon.

File: pkg/kubelet/eviction/helpers.go
Comment (ravisantoshgudimetla) [reply pkg/kubelet/eviction/helpers.go]: Or add a todo

File: pkg/controller/.import-restrictions
Comment (hprateek43) [reply pkg/controller/.import-restrictions]: I think we can remove this now post recent rebases. Will force push an update if local builds go through

Comment (ravisantoshgudimetla) [reply pkg/controller/.import-restrictions]: @deads2k - You're right, `k8s.io/kubernetes/pkg/controller/daemon` has a dependency on `pkg/scheduler/algorithm/predicates which in turn has a dependency on `k8s.io/kubernetes/pkg/scheduler/internal`(or the util before renaming). So, we should be ok with this but we need to see if we can refactor the predicates into a more neutral place that so both scheduler and DS controller can reuse them so that we can remove the dependecy

File: pkg/scheduler/internal/heap/heap.go
Comment (ravisantoshgudimetla) [reply pkg/scheduler/internal/heap/heap.go]: I agree, perhaps a TODO within the codebase?

File: pkg/controller/.import-restrictions
Comment (ravisantoshgudimetla) [reply pkg/controller/.import-restrictions]: /cc @bsalamat

**Review by ravisantoshgudimetla — COMMENTED** [review]
@hprateek43 - Thanks for the PR. I am fine with these changes, as long as you add some TODOs and create follow-up PRs on them
/lgtm

/hold

I want to give sometime to @deads2k and @misterikkit if they have any additional comments.

**Comment by deads2k** [comment]
Ok, since the internal dep is a transitive, I'll 

/approve

File: pkg/controller/.import-restrictions
Comment (liggitt) [reply pkg/controller/.import-restrictions]: > we should be ok with this but we need to see if we can refactor the predicates into a more neutral place that so both scheduler and DS controller can reuse them so that we can remove the dependency

we should do that *before* moving them into internal, rather than moving in a direction we don't want, then recovering

**Comment by hprateek43** [comment]
@misterikkit @bsalamat Can you review and approve this merge ?

**Comment by hprateek43** [comment]
@deads2k Can you review this now?

File: pkg/controller/.import-restrictions
Comment (misterikkit) [reply pkg/controller/.import-restrictions]: FYI I've discussed `.import-restrictions` with @lavalamp on past PRs and it's true that they list transitive dependencies. My opinion is that this is a quirk of the tool, and transitive dependencies on `pkg/scheduler/internal` are no worse than direct dependencies on `pkg/scheduler`

There are various k8s components that import bits of the scheduler code base, somewhat ad hoc. Moving code into our internal folder is a way to draw a line in the sand and prevent more packages from leaking into our supported API surface.

I also hope that getting explicit about which libraries are reasonable to use in other components will lead to better APIs and separation of concerns down the road.

File: pkg/kubelet/eviction/helpers.go
Comment (misterikkit) [inline pkg/kubelet/eviction/helpers.go]: Did this check come in because of a rebase?

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (misterikkit) [inline pkg/scheduler/internal/queue/scheduling_queue.go]: these import aliases are redundant. If your IDE added them, then I don't want to spend effort on it. If you're manually adding them, I think we should remove it.

**Review by misterikkit — COMMENTED** [review]
Looks like some of the files were moved from `pkg/scheduler/util` without changing the package name. I'm a little surprised the code compiles as-is.

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (hprateek43) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: My IDE added this redundant import. Do let me know if I should refactor it?

File: pkg/kubelet/eviction/helpers.go
Comment (hprateek43) [reply pkg/kubelet/eviction/helpers.go]: Yes. The recent rebase added this definition.

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (misterikkit) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: If your IDE added them, I don't want you (or anyone) to waste time manually fixing it. I think `goimports` somewhat recently changed its behavior around adding aliases. Ultimately, it's not doing any harm.

File: pkg/kubelet/eviction/helpers.go
Comment (misterikkit) [reply pkg/kubelet/eviction/helpers.go]: Hmm, if this check was added and you did a rebase, I would not expect it to show up as an added line in the diff.

It looks like what actually happened is that this check was removed in https://github.com/kubernetes/kubernetes/pull/79262 but is getting re-added in this PR. Could you remove it?

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (misterikkit) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: If your IDE added it, I don't want you (or anyone) wasting time manually fixing it. I think that `goimports` changed how it does import aliases a while back and it's causing things like this. In the end, it is not causing harm.

Comment (hprateek43) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: @misterikkit Ravi told me that he was awaiting some changes to be incorporated from @liggitt 's review. Do you have any suggestions of changes to be made into this PR ?

Comment (misterikkit) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: For import aliases, no changes are needed. For my other comment about the feature gate condition, that needs to change.

File: pkg/kubelet/eviction/helpers.go
Comment (hprateek43) [reply pkg/kubelet/eviction/helpers.go]: I have removed the feature gate. Please review now.

File: pkg/scheduler/internal/queue/scheduling_queue.go
Comment (hprateek43) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: I have removed that condition now.

**Comment by misterikkit** [comment]
/lgtm
I am happy with this. Thanks! :smile:

Need someone with approve powers to review.
@bsalamat @dashpole could you take a look?

**Review by misterikkit — COMMENTED** [review]
/lgtm

**Comment by misterikkit** [comment]
/lgtm cancel

Looks like build is failing after the rebase. @hprateek43 can you fix it up?

**Comment by hprateek43** [comment]
> /lgtm cancel
> 
> Looks like build is failing after the rebase. @hprateek43 can you fix it up?

I am working on this. Seems like some BUILD files have changed post this rebase

**Comment by hprateek43** [comment]
@misterikkit pull-kubernetes-integration has been a flaky test for me as most of the time some pod is getting timed out. Should I report this behaviour somewhere?

**Review by misterikkit — COMMENTED** [review]
/lgtm

**Comment by hprateek43** [comment]
@bowei @andrewsykim @bsalamat Can you review and approve this PR?

**Comment by hprateek43** [comment]
@derekwaynecarr @vishh @dchen1107 @dashpole @sjenning Can one of you review this PR and give an approval ?

**Comment by derekwaynecarr** [comment]
kubelet changes lgtm.

/approve
/lgtm

File: pkg/scheduler/internal/channels/error_channel.go
Comment (draveness) [inline pkg/scheduler/internal/channels/error_channel.go]: `channels` may not be a good name for package, `channel` would be better as a package name here.

File: pkg/scheduler/internal/heap/heap.go
Comment (draveness) [inline pkg/scheduler/internal/heap/heap.go]: does goimports add this automatically or it's a typo.

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (draveness) [inline pkg/scheduler/framework/v1alpha1/framework.go]: I think goimports only add alias when the package name is `v1`, it's the first time I saw this and looks weird.

File: pkg/scheduler/algorithm/priorities/interpod_affinity.go
Comment (draveness) [inline pkg/scheduler/algorithm/priorities/interpod_affinity.go]: what do you think of keeping the original name of the package? rename it to `schedutil` does not sound like a good idea.

File: pkg/scheduler/algorithm/predicates/metadata.go
Comment (draveness) [inline pkg/scheduler/algorithm/predicates/metadata.go]: This should be grouped with imports in L29-L37

**Review by draveness — CHANGES_REQUESTED** [review]

File: pkg/scheduler/core/extender_test.go
Comment (alculquicondor) [inline pkg/scheduler/core/extender_test.go]: It feels like SortableList shouldn't belong to the podinfo package.

**Review by alculquicondor — COMMENTED** [review]
My general feeling is that this PR can be broken up into at least 2.

For example:
PR1: move `Get*(v1.Pod)` into a new podinfo package
PR2: move heap into its own package
PR3: move SortableList into its own package.

arguably, heap and sortable list can live in the same package named. Arguably, heap and sortable list could be in the same package named "datastructures" or "containers".

File: pkg/scheduler/framework/v1alpha1/framework.go
Comment (Huang-Wei) [reply pkg/scheduler/framework/v1alpha1/framework.go]: +1, that's also my understanding. Please remove this unnecessary alias.

File: pkg/scheduler/core/extender_test.go
Comment (hprateek43) [reply pkg/scheduler/core/extender_test.go]: @alculquicondor Can you suggest a package to move it into?

File: pkg/scheduler/internal/channels/error_channel.go
Comment (hprateek43) [reply pkg/scheduler/internal/channels/error_channel.go]: @draveness @alculquicondor What package name would your suggest to go with?

Comment (draveness) [reply pkg/scheduler/internal/channels/error_channel.go]: I prefer `errchannel`.New` to `channel.NewErrorChannel`

Comment (alculquicondor) [reply pkg/scheduler/internal/channels/error_channel.go]: being more specific, as @draveness suggested, reads well too.

File: pkg/scheduler/algorithm/priorities/interpod_affinity.go
Comment (alculquicondor) [reply pkg/scheduler/algorithm/priorities/interpod_affinity.go]: +1

File: pkg/scheduler/core/extender_test.go
Comment (alculquicondor) [reply pkg/scheduler/core/extender_test.go]: what about `internal/collections` or more specific `internal/sortablelist`

**Comment by hprateek43** [comment]
@alculquicondor @draveness Can you suggest the right PR titles to split this PR?

**Comment by alculquicondor** [comment]
Your call. If something is very off, I will request a change.

**Comment by alculquicondor** [comment]
/close
This is quite stale now.
Perhaps the pod utils can be moved to a new package once #90039 lands

**Pull request closed without merging**
